File: app/src/main/java/space/pal/sig/service/download/DataSyncManager.java
package space.pal.sig.service.download;

import android.content.Context;

import androidx.annotation.NonNull;
import androidx.work.Worker;
import androidx.work.WorkerParameters;

import java.io.IOException;

import space.pal.sig.Space;
import space.pal.sig.injection.ApplicationComponent;
import space.pal.sig.repository.ApodRepository;
import space.pal.sig.repository.LaunchesRepository;
import space.pal.sig.repository.NewsRepository;
import space.pal.sig.repository.service.HubbleService;
import space.pal.sig.repository.service.LaunchService;
import space.pal.sig.repository.service.NasaService;

import static space.pal.sig.service.download.DataSyncUtil.downloadApod;
import static space.pal.sig.service.download.DataSyncUtil.downloadEsaNews;
import static space.pal.sig.service.download.DataSyncUtil.downloadHubbleNews;
import static space.pal.sig.service.download.DataSyncUtil.downloadJwNews;
import static space.pal.sig.service.download.DataSyncUtil.downloadNextLaunches;
import static space.pal.sig.service.download.DataSyncUtil.downloadStNews;
import static space.pal.sig.service.notification.SpaceNotificationManager.createNotification;

/**
 * SpaceNow
 * Created by Catalin on 7/31/2020
 **/
public class DataSyncManager extends Worker {

    public static final String DATA_SYNC_WORKER_TAG = "data_sync_worker_tag";

    public DataSyncManager(@NonNull Context context,
                           @NonNull WorkerParameters workerParams) {
        super(context, workerParams);
    }

    @NonNull
    @Override
    public Result doWork() {
        ApplicationComponent component = Space.getApplicationComponent();
        NasaService nasaService = component.nasaService();
        ApodRepository apodRepository = component.apodRepository();
        HubbleService hubbleService = component.hubbleService();
        NewsRepository newsRepository = component.newsRepository();
        LaunchService launchService = component.launchService();
        LaunchesRepository launchesRepository = component.launchesRepository();
        try {
            downloadApod(nasaService, apodRepository);
            downloadEsaNews(hubbleService, newsRepository);
            downloadHubbleNews(hubbleService, newsRepository);
            downloadJwNews(hubbleService, newsRepository);
            downloadStNews(hubbleService, newsRepository);
            downloadNextLaunches(launchService, launchesRepository);
            createNotification(getApplicationContext(), "Sync finished successfully.");
        } catch (IOException e) {
            e.printStackTrace();
            return Result.failure();
        }
        return Result.success();
    }
}

File: app/src/main/java/space/pal/sig/service/download/DataSyncManager.kt
package space.pal.sig.service.download

import android.content.Context
import androidx.work.Worker
import androidx.work.WorkerParameters
import com.google.gson.Gson
import org.koin.core.component.KoinComponent
import org.koin.core.component.inject
import space.pal.sig.R
import space.pal.sig.model.dto.FactDto
import space.pal.sig.model.entity.NewsSource
import space.pal.sig.repository.*

/**
 * SpaceNow
 * Created by Catalin on 7/31/2020
 */
class DataSyncManager(
        context: Context,
        workerParams: WorkerParameters
) : Worker(context, workerParams), KoinComponent {

    private val launchRepository: LaunchRepository by inject()
    private val roadsterRepository: RoadsterRepository by inject()
    private val apodRepository: ApodRepository by inject()
    private val rocketRepository: RocketRepository by inject()
    private val launchPadRepository: LaunchPadRepository by inject()
    private val factRepository: FactRepository by inject()
    private val newsRepository: NewsRepository by inject()
    private val gson = Gson()

    override fun doWork(): Result {
        syncAstronomyPicturesOfTheDay()
        syncRoadster()
        syncLaunches()
        syncRockets()
        syncLaunchpads()
        syncFacts()
        syncNews()
        return Result.success()
    }

    private fun syncAstronomyPicturesOfTheDay() {
        val apod = apodRepository
                .downloadCurrent()
                .blockingGet()
        apodRepository.save(apod.toAstronomyPictureOfTheDay())
    }

    private fun syncRoadster() {
        val roadsterDto = roadsterRepository
                .downloadRoadster()
                .blockingGet()
        roadsterRepository.save(roadsterDto.toRoadster())
    }

    private fun syncLaunches() {
        val launches = launchRepository
                .downloadAllLaunches()
                .blockingGet()
        launches.forEach { launchDto ->
            launchRepository.save(launchDto.toLaunch())
        }
    }

    private fun syncRockets() {
        val rockets = rocketRepository
                .downloadAllRockets()
                .blockingGet()
        rockets.forEach { rocketDto ->
            rocketRepository.save(rocketDto.toRocket())
        }
    }

    private fun syncLaunchpads() {
        val launchpads = launchPadRepository
                .downloadAllLaunchpads()
                .blockingGet()
        launchpads.forEach { launchpadDto ->
            launchPadRepository.save(launchpadDto.toLaunchpad())
        }
    }

    private fun syncFacts() {
        val factsJson: String = DataSyncUtil.readJsonFromResource(R.raw.facts, applicationContext)
        val facts = gson.fromJson(factsJson, Array<FactDto>::class.java)
        facts.forEach { factDto -> factRepository.save(factDto.toFact()) }
    }

    private fun syncNews() {
        for (i in 1..3) {
            val hubblePreviewNews = newsRepository
                    .downloadHubbleFeed(i)
                    .blockingGet()
            hubblePreviewNews.forEach { newsPreviewDto ->
                val newsDto = newsRepository
                        .downloadHubbleNews(newsPreviewDto.id)
                        .blockingGet()
                newsRepository.save(newsDto.toNews(NewsSource.HUBBLE))
            }
            val esaNews = newsRepository
                    .downloadEuropeanSpaceAgencyFeed(i)
                    .blockingGet()
            esaNews.forEach { newsDto ->
                newsRepository.save(newsDto.toNews(NewsSource.EUROPEAN_SPACE_AGENCY))
            }
            val jwstNews = newsRepository
                    .downloadJamesWebbSpaceTelescopeFeed(i)
                    .blockingGet()
            jwstNews.forEach { newsDto ->
                newsRepository.save(newsDto.toNews(NewsSource.JAMES_WEBB_SPACE_TELESCOPE))
            }
        }
        val stNews = newsRepository
                .downloadSpaceTelescopeLiveFeed(1)
                .blockingGet()
        stNews.forEach { newsDto ->
            newsRepository.save(newsDto.toNews(NewsSource.SPACE_TELESCOPE_LIVE))
        }
    }

    companion object {
        const val DATA_SYNC_WORKER_TAG = "data_sync_worker_tag"
    }
}